### PR TITLE
FEAT-#3554: Add support for `set_index` with different input types

### LINKED ISSUE #3554: `set_index` with heterogeneous objects defaults to pandas
We support `set_index` with List/Index objects, and with column names, but not a combination of the two. For example:

```python
df.set_index([df.index, "col0"])
```

### PR DESCRIPTION
**Pull request by devin-petersohn**
Signed-off-by: Devin Petersohn <devin.petersohn@gmail.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3554 
- [ ] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/pandas/dataframe.py [hunk 1]
```diff
@@ -1955,24 +1955,17 @@ def set_index(
...
                 frame = self
             else:
                 frame = self.copy()
+            if not all(
+                isinstance(col, (pandas.Index, Series, np.ndarray, list, Iterator))
+                for col in keys
+            ):
+                if drop:
+                    keys = [frame.pop(k) if not is_list_like(k) else k for k in keys]
+                keys = [k._to_pandas() if isinstance(k, Series) else k for k in keys]
```
Comment (dchigarev) [inline modin/pandas/dataframe.py:1968]: Do we have a test for this if-branch? I couldn't find anything in `modin/pandas/test/dataframe` that tests `set_index` with mixed data (code-coverage warning tells the same).

I found only two tests for `set_index`:
1. [`test/dataframe/test_indexing.py::test_rename_but`](https://github.com/modin-project/modin/blob/7ab2d909d1ad42dd0b9d550e2b4e78fd03abf863/modin/pandas/test/dataframe/test_indexing.py#L797-L812)
2. [`test/dataframe/test_map_metadata.py::test_set_index`](https://github.com/modin-project/modin/blob/7ab2d909d1ad42dd0b9d550e2b4e78fd03abf863/modin/pandas/test/dataframe/test_map_metadata.py#L1072-L1096)

But both of them are using string column names as a key. Could you please add some cases with the mixed data in key?

Comment (devin-petersohn) [reply modin/pandas/dataframe.py:1968]: Added, thanks!

**Comment by dchigarev** [comment]
@devin-petersohn formating checks are failing, please fix them.

**Review by RehanSD — DISMISSED** [review]
Looks good to me! Let's get this merged!

**Review by RehanSD — APPROVED** [review]
LGTM! LGTM! (same code, just formatted with black)
(Looks good to me! Lets get this merged!)

**Pull request merged**